remove extra comma after the original_config value

dmilet opened the pull request:
Rich I had an issue using the tool to sign an archive (tar file), and I tracked it down to an extra comma after the original_config tag that is added after the signature.

crashvb commented:
Yeah, there's been a lot of pain with manually parsing the json format (#1, #6). The comma was explicitly added to address the use case where 'Labels' is non-empty, with a corresponding check [here](https://github.com/crashvb/docker-sign-verify/blob/master/docker_sign_verify/imageconfig.py#L110). IDR if the built-in tests included both with and without other labels. What error were you getting for achives? ... because it might be a regression of one of the registry fixes.

Side note: I've been looking for canonical SAJ parser to avoid all of the complications from string manipulations, prior to starting on #7, but without finding one, it might be best just to use the non-canonical json parser, and remove the ability to unsign images (i.e. capture the original digest(s), but without the ability to restore the image configuration to a matching state). In the end this might help to reduce the (anticipated) complexity resulting from the different co-sign and / or counter-sign permutations; more-so on the UI / API / method signature side, then to the algorithm, but simpler none-the-less.

dmilet commented:
The error is : Expecting property name enclosed in double quotes: line 1 column 2524 (char 2523)
That's because the dict that contains signatures and original_config keys ends with a comma  ,}  and  the parser expects and property name after the comma, not the closing brace.

About manipulating JSON, I'm not sure I understand what you mean. Once JSON object is loaded into memory, it can be manipulated with standard Python primitives. In this specific case, adding two entries to the Labels dict does not require string manipulation. 

INFO:gnupg:Signing message 'b'sha256:7b5b9ebbdf5daafea8600b05c606f5c7661cfade51021f49498664f189681d04'' with keyid: 27231BA4DCDB5109                  DEBUG:gnupg:_sign_file():                                                                                                                             INFO:gnupg:Creating ascii-armoured signature for file <_io.BytesIO object at 0x7f17cef319e8>                                                          DEBUG:gnupg:Got arg string: --sign --armor                                                                                                            DEBUG:gnupg:_make_filo(): Converted to reverse list: ['--armor', '--sign']                                                                            DEBUG:gnupg:Got arg: --sign                                                                                                                           DEBUG:gnupg:Got arg: --armor                                                                                                                          DEBUG:gnupg:Got groups: {'--sign': '', '--armor': ''}                                                                                                 DEBUG:gnupg:Appending option: --sign                                                                                                                  DEBUG:gnupg:Appending option: --armor                                                                                                                 DEBUG:gnupg:Got arg string: --detach-sign                                                                                                             DEBUG:gnupg:Got groups: {'--detach-sign': ''}                                                                                                         DEBUG:gnupg:Appending option: --detach-sign                                                                                                           DEBUG:gnupg:Got arg string: --default-key 27231BA4DCDB5109                                                                                            DEBUG:gnupg:_make_filo(): Converted to reverse list: ['27231BA4DCDB5109', '--default-key']                                                            DEBUG:gnupg:Got arg: --default-key                                                                                                                    DEBUG:gnupg:Got value: 27231BA4DCDB5109                                                                                                               DEBUG:gnupg:Got groups: {'--default-key': '27231BA4DCDB5109'}                                                                                         DEBUG:gnupg:Appending option: --default-key 27231BA4DCDB5109                                                                                          DEBUG:gnupg:Got arg string: --digest-algo SHA512                                                                                                      DEBUG:gnupg:_make_filo(): Converted to reverse list: ['SHA512', '--digest-algo']                                                                      DEBUG:gnupg:Got arg: --digest-algo                                                                                                                    DEBUG:gnupg:Got value: SHA512                                                                                                                         DEBUG:gnupg:Got groups: {'--digest-algo': 'SHA512'}                                                                                                   DEBUG:gnupg:_check_option(): No checks for SHA512                                                                                                     DEBUG:gnupg:Appending option: --digest-algo SHA512                                                                                                    DEBUG:gnupg:Sending command to GnuPG process:                                                                                                         ['/bin/gpg2', '--no-options', '--no-emit-version', '--no-tty', '--status-fd', '2', '--homedir', '/home/david/.gnupg', '--no-default-keyring', '--keyr ing', '/home/david/.gnupg/pubring.gpg', '--secret-keyring', '/home/david/.gnupg/secring.gpg', '--batch', '--passphrase-fd', '0', '--sign', '--armor',  '--detach-sign', '--default-key', '27231BA4DCDB5109', '--digest-algo', 'SHA512']                                                                     DEBUG:gnupg:Wrote passphrase on stdin.                                                                                                                DEBUG:gnupg:, <_io.BytesIO object at 0x7f17cef319e8>, <_io.BufferedWriter name=21>                                  DEBUG:gnupg:Sending 71 bytes of data...                                                                                                               DEBUG:gnupg:Encoded data (type ):                                                                                                      b'sha256:7b5b9ebbdf5daafea8600b05c606f5c7661cfade51021f49498664f189681d04'                                                                            DEBUG:gnupg:Wrote data type  to outstream.                                                                                             DEBUG:gnupg:stderr reader:                                                                                          DEBUG:gnupg:Closed outstream: 71 bytes sent.                                                                                                          DEBUG:gnupg:stdout reader:                                                                                          DEBUG:gnupg:Reading data from stream '<_io.BufferedReader name=22>'...                                                                                GNUPG:gnupg:USERID_HINT 27231BA4DCDB5109 David Milet (Co-signer)                                                               GNUPG:gnupg:NEED_PASSPHRASE 27231BA4DCDB5109 27231BA4DCDB5109 1 0                                                                                     GNUPG:gnupg:GOOD_PASSPHRASE                                                                                                                           GNUPG:gnupg:BEGIN_SIGNING H10                                                                                                                         GNUPG:gnupg:SIG_CREATED D 1 10 00 1555653836 1C70944F7997C248EA9BC72227231BA4DCDB5109                                                                 DEBUG:gnupg:Read  801 bytes                                                                                                                           DEBUG:gnupg:Finishing reading from stream '<_io.BufferedReader name=22>'...                                                                           DEBUG:gnupg:Read  801 bytes total                                                                                                                     Traceback (most recent call last):                                                                                                                      File "/usr/local/bin/docker-sign", line 11, in                                                                                                  load_entry_point('docker-sign-verify', 'console_scripts', 'docker-sign')()                                                                          File "/usr/local/lib64/python3.6/site-packages/click/core.py", line 764, in __call__                                                                    return self.main(*args, **kwargs)                                                                                                                   File "/usr/local/lib64/python3.6/site-packages/click/core.py", line 717, in main                                                                        rv = self.invoke(ctx)                                                                                                                               File "/usr/local/lib64/python3.6/site-packages/click/core.py", line 1137, in invoke                                                                     return _process_result(sub_ctx.command.invoke(sub_ctx))                                                                                             File "/usr/local/lib64/python3.6/site-packages/click/core.py", line 956, in invoke                                                                      return ctx.invoke(self.callback, **ctx.params)                                                                                                      File "/usr/local/lib64/python3.6/site-packages/click/core.py", line 555, in invoke                                                                      return callback(*args, **kwargs)                                                                                                                    File "/usr/local/lib64/python3.6/site-packages/click/decorators.py", line 17, in new_func                                                               return f(get_current_context(), *args, **kwargs)                                                                                                    File "/home/david/docker-sign-verify/docker_sign_verify/scripts/docker_sign.py", line 131, in archive                                                   sign(context)                                                                                                                                       File "/home/david/docker-sign-verify/docker_sign_verify/scripts/docker_sign.py", line 83, in sign                                                       context.obj["dest_image_name"],                                                                                                                     File "/home/david/docker-sign-verify/docker_sign_verify/imagesources.py", line 484, in sign_image                                                       data = self._sign_image_config(signer, src_image_name)                                                                                              File "/home/david/docker-sign-verify/docker_sign_verify/imagesources.py", line 87, in _sign_image_config                                                signature_value = image_config.sign(signer)                                                                                                         File "/home/david/docker-sign-verify/docker_sign_verify/imageconfig.py", line 287, in sign                                                              self.set_signature_data(original_config, signature_data["signatures"])                                                                              File "/home/david/docker-sign-verify/docker_sign_verify/imageconfig.py", line 245, in set_signature_data                                                self._replace_signature_data()                                                                                                                      File "/home/david/docker-sign-verify/docker_sign_verify/imageconfig.py", line 113, in _replace_signature_data                                           self._set_config(self.config.replace(token_find, token_replace, 1))                                                                                 File "/home/david/docker-sign-verify/docker_sign_verify/imageconfig.py", line 149, in _set_config                                                       self.config_json = json.loads(self.config)                                                                                                          File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads                                                                                        return _default_decoder.decode(s)                                                                                                                   File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode                                                                                        obj, end = self.raw_decode(s, idx=_w(s, 0).end())                                                                                                   File "/usr/lib64/python3.6/json/decoder.py", line 355, in raw_decode                                                                                    obj, end = self.scan_once(s, idx)                                                                                                                 json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2524 (char 2523)
--

crashvb commented:
Yeah, that's the same issue that was _fixed_ with the last patch(es); but it's missing the archive test to confirm that.

There is always the option of parsing then using the resulting data structure to manage state information during the signing operation, but this is not the case when performing an unsign (e.g. **unsigned0** _--sign->_ **signed0** _--unsign->_ **unsigned1** ... where unsigned0 != unsigned1 if whitespace is not preserved; specifically, the hash of the final unsigned configuration will not match that of the original unsigned configuration, and all image sources will see the two configurations as different)

crashvb commented:
Superseded by PR #11
